packages/flutter/lib/src/rendering/mouse_tracking.dart

```diff
@@ -397,71 +670,146 @@ class MouseTracker extends ChangeNotifier {
       // Otherwise, trigger the hover callback only on annotations that it
       // newly enters.
       final Iterable<MouseTrackerAnnotation> hoveringAnnotations = pointerHasMoved ? nextAnnotations.toList().reversed : enteringAnnotations;
       for (final MouseTrackerAnnotation annotation in hoveringAnnotations) {
         if (annotation.onHover != null) {
-          annotation.onHover(unhandledEvent);
+          annotation.onHover(triggeringEvent);
         }
       }
     }
   }
 
-  bool _hasScheduledPostFrameCheck = false;
-  /// Mark all devices as dirty, and schedule a callback that is executed in the
-  /// upcoming post-frame phase to check their updates.
-  ///
-  /// Checking a device means to collect the annotations that the pointer
-  /// hovers, and triggers necessary callbacks accordingly.
-  ///
-  /// Although the actual callback belongs to the scheduler's post-frame phase,
-  /// this method must be called in persistent callback phase to ensure that
-  /// the callback is scheduled after every frame, since every frame can change
-  /// the position of annotations. Typically the method is called by
-  /// [RendererBinding]'s drawing method.
-  void schedulePostFrameCheck() {
-    assert(_duringBuildPhase);
-    assert(!_duringDeviceUpdate);
-    if (!mouseIsConnected)
+  /// MOUSE CURSOR
+
+  /// Returns the current mouse cursor of a device.
+  /// 
+  /// Only valid when asserts are enabled. In release builds, always returns
+  /// null.
+  PreparedMouseCursor debugDeviceCursor(int device) {
+    PreparedMouseCursor result;
+    assert(() {
+      result = _mouseCursorStates[device]?.currentAnnotation?.cursor;
+      return true;
+    }());
+    return result;
+  }
+  final Map<int, _MouseCursorState> _mouseCursorStates = <int, _MouseCursorState>{};
+
+  // Find the mouse cursor, which fallbacks to _fallbackAnnotation.
+  // The `annotations` is the current annotations that the device is
+  // hovering in visual order from front the back.
+  // The return value is never null.
+  MouseTrackerAnnotation _findCursorAnnotation(LinkedHashSet<MouseTrackerAnnotation> annotations) {
+    MouseTrackerAnnotation result;
+    for (final MouseTrackerAnnotation annotation in annotations) {
+      if (annotation.cursor != null) {
+        result = annotation;
+      }
+    }
+    return result ?? const _FallbackAnnotation();
+  }
+
+  // Define the actions when a pointer should change to a prepared cursor by
+  // calling system functions.
+  Future<void> _handleActivateCursor(int device, PreparedMouseCursor cursor) async {
+    if (cursor is NoopMouseCursor)
       return;
-    if (!_hasScheduledPostFrameCheck) {
-      _hasScheduledPostFrameCheck = true;
-      SchedulerBinding.instance.addPostFrameCallback((Duration duration) {
-        assert(_hasScheduledPostFrameCheck);
-        _hasScheduledPostFrameCheck = false;
-        _updateAllDevices();
-      });
+
+    if (cursor is SystemMouseCursor) {
```

Comment:
We generally try to avoid "is" type checks.

Reply:
Actually I see a ton of `is` type checks everywhere... I'm ok with changing to `runtimeType` if you insist though. Everything is hard-coded here anyway.

Reply:
Could each MouseCursor just know how to activate itself and you just call `cursor.activate()` here?

Reply:
That was actually an [earlier design](https://github.com/flutter/flutter/pull/42797/files#diff-4acb3f1c87ad466213ead70a0a138093R189).

As far as I can remember, the biggest reason we didn't like it is because it lured developers into imperative calls, since they can now write
```dart
MouseRegion(
  onEnter: (_) { myCursor.activate(); }
);
```

Reply:
I still don't like the type check as it braids two concepts together that seem separate, see also https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-complecting-braiding-multiple-concepts-together.

Why does the Mouse**Tracker** need to know how to activate a cursor?

Reply:
It's not so much `is` checks that we avoid, it's having the management class know about the subservient classes. To put it another way: how would someone implement their own version of SystemMouseCursor, or NoopMouseCursor? If you hard-code knowledge of those classes here, it's not possible.

Instead we define a small protocol for the management class to talk to the subservient class and do everything through that, without the management class needing to know about anything. That's essentially how we've built up our composition approach everywhere.

Reply:
Changed as described in https://github.com/flutter/flutter/pull/54171#issuecomment-617535039